src/core/dbus-unit.c

```diff
@@ -857,6 +857,39 @@ static int property_get_refs(
         return sd_bus_message_close_container(reply);
 }
 
+static int property_get_activation_event_info(
+                sd_bus *bus,
+                const char *path,
+                const char *interface,
+                const char *property,
+                sd_bus_message *reply,
+                void *userdata,
+                sd_bus_error *error) {
+
+        _cleanup_strv_free_ char **pairs = NULL;
+        Unit *u = ASSERT_PTR(userdata);
+        int r;
+
+        assert(bus);
+        assert(reply);
+
+        r = activation_event_info_append_pair(u->activation_event_info, &pairs);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_open_container(reply, 'a', "(ss)");
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH_PAIR(key, value, pairs) {
+                r = sd_bus_message_append(reply, "(ss)", *key, *value);
+                if (r < 0)
+                        return r;
+        }
+
+        return sd_bus_message_close_container(reply);
```

Comment:
dbus-job.c has pretty much the same function now, even under the same name. Please add a helper somewhere that unifies the main part of this function somewhere

Reply:
also, there's sd_Bus_message_append_strv()

Reply:
sd_bus_message_append_strv appends as `as`, but here I need key/value pair, `a(ss)`